<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • bcbioBase</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Changelog" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioBase</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioBase">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/hbc/bcbioBase/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="bcbiobase-0-4-0-2018-08-08" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-4-0-2018-08-08" class="anchor"></a>bcbioBase 0.4.0 (2018-08-08)</h2>
<p>This is a maintenance release designed to simplify the package for long-term stability. Here we are moving all of the current S4 method support to <a href="http://steinbaugh.com/basejump">basejump</a> for consistency, since the methods apply to <code>SummarizedExperiment</code> and are not <a href="http://bcb.io">bcbio</a>-workflow specific. We are keeping the package simple by exporting the <code>read*</code> family of functions here, which are designed to integrate with the <a href="http://bcb.io">bcbio</a> output directories.</p>
<div id="major-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h3>
<ul>
<li>Moved S4 method support to <a href="http://steinbaugh.com/basejump">basejump</a>: <code>flatFiles</code>, <code>metrics</code>, <code>plotCorrelationHeatmap</code>, <code>plotDot</code>, <code>plotGene</code>, <code>plotHeatmap</code>, <code>plotQuantileHeatmap</code>, <code>plotViolin</code>.</li>
<li>
<code>prepareSummarizedExperiment</code> has been moved to <a href="http://steinbaugh.com/basejump">basejump</a> and renamed to <code>makeSummarizedExperiment</code>, for consistency. The underlying code remains the same.</li>
<li>Moved the <code>SummarizedExperiment</code> to <code>list</code> coercion method to <a href="http://steinbaugh.com/basejump">basejump</a>.</li>
<li>Reduced the number of re-exported <a href="https://ggplot2.tidyverse.org">ggplot2</a> and <a href="https://cran.r-project.org/package=viridis">viridis</a> functions, for simplicity.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-3-2-2018-07-31" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-3-2-2018-07-31" class="anchor"></a>bcbioBase 0.3.2 (2018-07-31)</h2>
<div id="minor-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes" class="anchor"></a>Minor changes</h3>
<ul>
<li>Offloaded <code>separatorBar</code> and <code>updateMsg</code> to <a href="http://steinbaugh.com/basejump">basejump</a>.</li>
<li>Now using <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> v6.1 for documentation.</li>
<li>Miscellaneous documentation improvements.</li>
<li>Simplified installation instructions, referring the user to <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> and <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a> packages instead.</li>
<li>
<a href="https://github.com/jimhester/lintr">lintr</a> check fixes.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-3-1-2018-07-24" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-3-1-2018-07-24" class="anchor"></a>bcbioBase 0.3.1 (2018-07-24)</h2>
<div id="minor-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h3>
<ul>
<li>Added hexadecimal color support to heatmap functions.</li>
<li>Improved documentation, specifying the supported types for each argument more clearly.</li>
<li>Improved messages in <code>readYAML</code> family of functions.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-3-0-2018-07-17" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-3-0-2018-07-17" class="anchor"></a>bcbioBase 0.3.0 (2018-07-17)</h2>
<div id="major-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h3>
<ul>
<li>Moved some generics that provide <code>SummarizedExperiment</code> method suport to <a href="http://steinbaugh.com/basejump">basejump</a> package: <code>gene2symbol</code>, <code>interestingGroups</code>, <code>sampleData</code>, <code>sampleNames</code>, <code>sanitizeSampleData</code>, <code>selectSamples</code>, <code>uniteInterestingGroups</code>.</li>
<li>Moved method support for <code>SummarizedExperiment</code> to <a href="http://steinbaugh.com/basejump">basejump</a>: <code>convertGenesToSymbols</code>, <code>counts</code>, <code>gene2symbol</code>, <code>interestingGroups</code>, <code>sampleData</code>, etc.</li>
<li>Heatmap functions now only provide <code>SummarizedExperiment</code> method support. The documentation for these heatmap functions has been improved (and simplified), showing the supported arguments more clearly.</li>
</ul>
</div>
<div id="minor-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h3>
<ul>
<li>Now using <code>transcriptID</code> instead of <code>txID</code> in tx2gene and GRanges metadata defined in <code>rowRanges</code> slot of object.</li>
<li>Now using <code>aes</code> instead of <code>aes_string</code> for all internal <a href="https://ggplot2.tidyverse.org">ggplot2</a> code, which uses tidyeval (v3.0 update).</li>
<li>Moved <code>rse_bcb</code> and <code>rse_bcb</code> example data to <a href="http://steinbaugh.com/basejump">basejump</a>, for unit testing. These datasets are still available in <a href="http://bioinformatics.sph.harvard.edu/bcbioBase">bcbioBase</a>.</li>
<li>Made <code>assertFormalAnnotationCol</code> function defunct. No longer using to check <code>annotationCol</code> argument integration in heatmap functions, since we’re now providing only <code>SummarizedExperiment</code> method support, instead of matrix method support (which is too complicated and error prone for users).</li>
<li>Moved <code>assertFormalInterestingGroups</code> function to <a href="http://steinbaugh.com/basejump">basejump</a>, where the other assert check functions are defined. This helps improve package consistency.</li>
<li>Updated unit tests to reflect changes in the number of exported generics.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-16-2018-06-28" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-16-2018-06-28" class="anchor"></a>bcbioBase 0.2.16 (2018-06-28)</h2>
<div id="minor-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-3" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved aggregate column handling in <code>bcbio_geom_abline</code>.</li>
<li>Improved code coverage back to 100%.</li>
<li>Reorganized R methods files to use <code>-methods.R</code> as a suffix, as recommended by <a href="http://bioconductor.org">Bioconductor</a>, in preparation for package submission.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-15-2018-06-05" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-15-2018-06-05" class="anchor"></a>bcbioBase 0.2.15 (2018-06-05)</h2>
<div id="new-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>minimalSampleData</code> enables easy creation of a sample metadata <code>data.frame</code> by simplify specifying the sample names (e.g. “description” in <a href="http://bcb.io">bcbio</a> YAML). This function was added for easy metadata handling for 10X Cell Ranger in the <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a> package, but is generally applicable to other <a href="http://bcb.io">bcbio</a> datasets and may be incorporated into <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> in a future update.</li>
</ul>
</div>
<div id="minor-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-4" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved internal <code>interestingGroups</code> handling inside <code>plotHeatmap</code> family of functions.</li>
<li>Removed message about transgeneNames and spikeNames in <code>prepareSummarizedExperiment</code> if missing transcripts are present. Now the function simply lists the genes that don’t have metadata in <code>rowRanges</code>.</li>
<li>Reorganized assert checks imports in <code>bcbioBase-package.R</code> file.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-14-2018-05-18" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-14-2018-05-18" class="anchor"></a>bcbioBase 0.2.14 (2018-05-18)</h2>
<div id="minor-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved formatting of package NEWS.</li>
<li>Improved code coverage by adding ggplot2 unit testing.</li>
<li>Added <code>sampleData&lt;-</code> method support for standard <code>data.frame</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-13-2018-05-07" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-13-2018-05-07" class="anchor"></a>bcbioBase 0.2.13 (2018-05-07)</h2>
<div id="minor-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h3>
<ul>
<li>Added <code>overwrite</code> parameter to <code>prepareTemplate</code>, disabled by default.</li>
<li>Removed validity check for <code>sampleData</code> accessor.</li>
<li>Improved vector return in <code>sampleNames</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-12-2018-05-08" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-12-2018-05-08" class="anchor"></a>bcbioBase 0.2.12 (2018-05-08)</h2>
<div id="minor-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated <code>sanitizeSampleData</code> from basejump to bcbioBase.</li>
<li>Don’t return <code>interestingGroups</code> column in <code>sampleData</code> return when <code>clean = TRUE</code>.</li>
<li>Removed internal R Markdown shared files. These are already saved per bcbio R package.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-11-2018-05-07" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-11-2018-05-07" class="anchor"></a>bcbioBase 0.2.11 (2018-05-07)</h2>
<div id="new-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#new-methods" class="anchor"></a>New methods</h3>
<ul>
<li>Added <code>SummarizedExperiment</code> method support for <code>sampleNames</code> generic.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-10-2018-05-03" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-10-2018-05-03" class="anchor"></a>bcbioBase 0.2.10 (2018-05-03)</h2>
<div id="new-functions-1" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-1" class="anchor"></a>New functions</h3>
<ul>
<li>Exporting <a href="https://ggplot2.tidyverse.org">ggplot2</a> convenience functions for easier plotting in <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> and <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a> packages: <code>bcbio_geom_abline</code>, <code>bcbio_geom_label</code>, <code>bcbio_geom_label_average</code>, and <code>bcbio_geom_label_repel</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-9-2018-04-30" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-9-2018-04-30" class="anchor"></a>bcbioBase 0.2.9 (2018-04-30)</h2>
<div id="minor-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved assert check messages for interesting groups.</li>
<li>Switched back to internally using <code>make.unique</code> instead of <code>make.names</code> for <code>convertGenesToSymbols</code> coercion on SummarizedExperiment objects. We’re using <code>make.names</code> only when coercing seurat objects (see <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a> code).</li>
<li>Updated <code>RangedSummarizedExperiment</code> working example, based on <code>bcbioRNASeq</code> example dataset.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-8-2018-04-25" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-8-2018-04-25" class="anchor"></a>bcbioBase 0.2.8 (2018-04-25)</h2>
<div id="minor-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-9" class="anchor"></a>Minor changes</h3>
<ul>
<li>Split out assertive imports so we can pin release to <a href="https://bioconda.github.io">bioconda</a>.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-7-2018-04-25" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-7-2018-04-25" class="anchor"></a>bcbioBase 0.2.7 (2018-04-25)</h2>
<div id="minor-changes-10" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-10" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated recommended Bioconductor installation method for 3.7 release.</li>
<li>Attempt to add <code>description</code> to <code>metadataBlacklist</code>.</li>
<li>
<code>readSampleData</code>: Use <code>description</code> metadata column internally instead of <code>sampleID</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-6-2018-04-25" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-6-2018-04-25" class="anchor"></a>bcbioBase 0.2.6 (2018-04-25)</h2>
<div id="minor-changes-11" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-11" class="anchor"></a>Minor changes</h3>
<ul>
<li>Using <code>make.names</code> instead of <code>make.unique</code> in <code>convertGenesToSymbols</code> method for <code>SummarizedExperiment</code> class.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-5-2018-04-24" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-5-2018-04-24" class="anchor"></a>bcbioBase 0.2.5 (2018-04-24)</h2>
<div id="major-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>sampleData</code> now supports <code>clean = TRUE</code> argument, which will hide columns that contain metrics or other calculations from the user. This is useful when preparing an R Markdown report, where we only want to show relevant metadata (e.g. factor columns). This is enabled by default.</li>
<li>Now using <code>metadataBlacklist</code> to hide specific columns in sample metadata.</li>
<li>
<code>sampleMetadata</code> is now deprecated in favor of <code>sampleData</code>. This function still works but will now warn the user, and should be removed from R Markdown template code.</li>
<li>Added <code>SummarizedExperiment</code> to <code>list</code> coercion support, which uses the same code as <code>flatFiles</code>.</li>
<li>
<code>readSampleData</code> no longer requires or recommended <code>fileName</code> column. Only the <code>description</code> column is required for demultiplexed data. In the case of multiplexed samples (e.g. inDrops single-cell RNA-seq), then <code>sampleName</code>, <code>index</code> and <code>sequence</code> names are also required. Multiplexed cell ranger data only requires <code>sampleName</code> and <code>index</code>, since the index barcode isn’t present in the counts matrix.</li>
</ul>
</div>
<div id="minor-changes-12" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-12" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>sampleID</code> is no longer included in metadata priority columns. This is only to be used internally and should be hidden from the user where possible.</li>
<li>
<code>prepareSampleData</code> is no longer needed and is now defunct.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-4-2018-04-22" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-4-2018-04-22" class="anchor"></a>bcbioBase 0.2.4 (2018-04-22)</h2>
<div id="minor-changes-13" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-13" class="anchor"></a>Minor changes</h3>
<ul>
<li>Reworked internal code for <code>readYAML</code> family of functions.</li>
<li>Reorganized and improved default arguments for heatmap functions.</li>
<li>
<code>readSampleData</code>: switched to using <code>merge</code> instead of <code>left_join</code> internally.</li>
<li>Reexporting viridis family of color functions, including <code>inferno</code>, and the British variant <code>scale_colour_viridis</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-3-2018-04-19" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-3-2018-04-19" class="anchor"></a>bcbioBase 0.2.3 (2018-04-19)</h2>
<ul>
<li>
<code>plotHeatmap</code>, <code>plotQuantileHeatmap</code>: Always attempt to convert genes to symbols for heatmaps.</li>
</ul>
</div>
    <div id="bcbiobase-0-2-2-2018-04-16" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-2-2018-04-16" class="anchor"></a>bcbioBase 0.2.2 (2018-04-16)</h2>
<div id="minor-changes-14" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-14" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved unit tests for <code>convertGenesToSymbols</code> and <code>gene2symbol</code>.</li>
<li>Miscellaneous documentation fixes.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-1-2018-04-13" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-1-2018-04-13" class="anchor"></a>bcbioBase 0.2.1 (2018-04-13)</h2>
<div id="major-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>convertGenesToSymbols</code> now has method support for <code>SummarizedExperiment</code>.</li>
<li>Updated example datasets to use <code>rse_bcb</code> and <code>rse_dds</code>.</li>
<li>Simplified internal S4 method code for heatmap functions.</li>
<li>
<code>prepareSummarizedExperiment</code> now supports transgenes with the <code>transgeneNames</code> argument (e.g. “EGFP”), and FASTA spike-ins with the <code>spikeNames</code> argument (e.g. “ERCCs”). Additionally, <code>rowRanges</code> and <code>colData</code> are no longer required and can be left <code>NULL</code>, although this isn’t generally recommended.</li>
</ul>
</div>
<div id="minor-changes-15" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-15" class="anchor"></a>Minor changes</h3>
<ul>
<li>Consolidated <code>sampleYAML</code> function code.</li>
<li>Consolidated assert check imports into <code>bcbioBase-package.R</code> file.</li>
<li>Consolidated globals into <code>globals.R</code> file.</li>
<li>Improved internal <a href="https://github.com/raivokolde/pheatmap">pheatmap</a> color support.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-2-0-2018-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-2-0-2018-03-22" class="anchor"></a>bcbioBase 0.2.0 (2018-03-22)</h2>
<div id="major-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-4" class="anchor"></a>Major changes</h3>
<ul>
<li>Added <code>sampleData</code> generic. Providing method support for <code>SummarizedExperiment</code> class. This generic provides consistent sample metadata support in both the <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> and <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a> packages. For <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a>, <code>colData</code> returns information about the cells, not the samples. For <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a>, <code>colData</code> and <code>sampleData</code> provide similar information on the samples.</li>
<li>Added <code>uniteInterestingGroups</code> as a generic. Provides support for adding an <code>interestingGroups</code> column to a data frame.</li>
<li>Added <code>SummarizedExperiment</code> method support for <code>gene2symbol</code>, <code>interestingGroups</code> generics.</li>
<li>Converted <code>prepareSummarizedExperiment</code> from a generic to a standard function.</li>
<li>
<code>prepareSummarizedExperiment</code> now requires rowRanges and supports FASTA spike-ins with <code>isSpike</code> argument. This helps set up a <code>SingleCellExperiment</code> object in the <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell">bcbioSingleCell</a> package.</li>
<li>
<code>prepareTemplate</code> has been converted from a generic to a standard function.</li>
<li>
<code>readDataVersions</code> and <code>readProgramVersions</code> now return an empty tibble if the file is missing. bcbio doesn’t always output these files, so we have changed the behavior from stopping on a missing file to simply warning and returning empty.</li>
<li>Added <code>sampleDirs</code> function, that informs the user about the names of the sample directories in the bcbio run.</li>
</ul>
</div>
<div id="minor-changes-16" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-16" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved internal factor sanitization of sample metadata for <code>readSampleMetadataFile</code> and <code>sampleYAMLMetadata</code>.</li>
<li>Exporting <code>projectDirPattern</code> to match bcbio project directories (e.g. <code>2018-01-01_rnaseq</code>) across packages.</li>
<li>Changed default of <code>rdsToken</code> in <code>copyToDropbox</code> to be <code>NULL</code> instead of <code>NA</code>.</li>
<li>Reduced the number of imports from basejump package.</li>
<li>Now using <code><a href="http://www.rdocumentation.org/packages/sessioninfo/topics/session_info">sessioninfo::session_info</a></code> instead of <code><a href="http://www.rdocumentation.org/packages/devtools/topics/session_info">devtools::session_info</a></code>.</li>
<li>Added AppVeyor CI support for code testing on Windows.</li>
<li>Made Travis CI checking stricter and added <code>BiocCheck</code>.</li>
<li>Simplified package dependencies in DESCRIPTION.</li>
</ul>
</div>
<div id="deprecations" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations" class="anchor"></a>Deprecations</h3>
<ul>
<li>Deprecated <code>sampleMetadata</code> in favor of <code>sampleData</code>.</li>
<li>Deprecated <code>flatFiles</code> in favor of using <code>as(object, "list")</code> coercion method instead.</li>
<li>Deprecated <code>prepareSampleMetadata</code> in favor of <code>prepareSampleData</code>.</li>
<li>Made <code>annotable</code> method on <code>SummarizedExperiment</code> objects defunct.</li>
</ul>
</div>
</div>
    <div id="bcbiobase-0-1-4-2018-02-20" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-1-4-2018-02-20" class="anchor"></a>bcbioBase 0.1.4 (2018-02-20)</h2>
<ul>
<li>Don’t include sample metadata in <code>summaryYAMLMetrics</code> return.</li>
</ul>
</div>
    <div id="bcbiobase-0-1-3-2018-02-19" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-1-3-2018-02-19" class="anchor"></a>bcbioBase 0.1.3 (2018-02-19)</h2>
<ul>
<li>Now exporting <code>assertFormalInterestingGroups</code> in camel case.</li>
</ul>
</div>
    <div id="bcbiobase-0-1-2-2018-02-18" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-1-2-2018-02-18" class="anchor"></a>bcbioBase 0.1.2 (2018-02-18)</h2>
<ul>
<li>Added back <code>checkInterestingGroups</code> since this code is still present on <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> master branch.</li>
</ul>
</div>
    <div id="bcbiobase-0-1-1-2018-02-17" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-1-1-2018-02-17" class="anchor"></a>bcbioBase 0.1.1 (2018-02-17)</h2>
<ul>
<li>Added internal assert checks.</li>
<li>Now exporting <code>sampleMetadata&lt;-</code> assignment generic.</li>
<li>Updated encrypted token for rdrop2 working example.</li>
<li>Added <code>assert_formal_interesting_groups</code> assert checks. Deprecated <code>checkInterestingGroups</code>.</li>
</ul>
</div>
    <div id="bcbiobase-0-1-0-2018-02-13" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-1-0-2018-02-13" class="anchor"></a>bcbioBase 0.1.0 (2018-02-13)</h2>
<ul>
<li>Preparing version pinning for <a href="https://bioconda.github.io">bioconda</a>. Relaxed <a href="http://rlang.tidyverse.org">rlang</a> dependency from v0.1.6 to v0.1.2 and <a href="http://tidyr.tidyverse.org">tidyr</a> dependency from v0.7.2 to v0.7.1.</li>
<li>Added <code>copyToDropbox</code> function, which enables input of a list of local files and returns <a href="https://www.dropbox.com">Dropbox</a> paths using <a href="https://github.com/karthik/rdrop2">rdrop2</a>.</li>
<li>Updated <a href="http://steinbaugh.com/basejump">basejump</a> dependency to v0.2.1.</li>
<li>Added assertive checks for all functions.</li>
</ul>
</div>
    <div id="bcbiobase-0-0-3-2018-01-27" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-0-3-2018-01-27" class="anchor"></a>bcbioBase 0.0.3 (2018-01-27)</h2>
<ul>
<li>NAMESPACE export fixes.</li>
<li>Minor code cleanup to pass <a href="https://github.com/jimhester/lintr">lintr</a> checks.</li>
<li>Switched to <a href="http://rlang.tidyverse.org">rlang</a> methods for errors, messages, and warnings: <code>abort</code>, <code>inform</code>, and <code>warn</code>.</li>
<li>Messages now consistently use backticks instead of apostrophes, as recommended by the <a href="http://style.tidyverse.org">tidyverse style guide</a>.</li>
<li>Removed <a href="http://tibble.tidyverse.org">tibble</a> rownames support from <code>prepareSampleMetadata</code>.</li>
<li>Improved code coverage to 100%.</li>
</ul>
</div>
    <div id="bcbiobase-0-0-2-2018-01-19" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-0-2-2018-01-19" class="anchor"></a>bcbioBase 0.0.2 (2018-01-19)</h2>
<ul>
<li>Re-export <code>assignAndSaveData</code>, <code>loadData</code>, <code>loadRemoteData</code>, <code>saveData</code> from <a href="http://steinbaugh.com/basejump">basejump</a> package.</li>
<li>Update <code>prepareTemplate</code> function to use internally stored data.</li>
</ul>
</div>
    <div id="bcbiobase-0-0-1-2018-01-11" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiobase-0-0-1-2018-01-11" class="anchor"></a>bcbioBase 0.0.1 (2018-01-11)</h2>
<ul>
<li>Initial release.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiobase-0-4-0-2018-08-08">0.4.0</a></li>
        <li><a href="#bcbiobase-0-3-2-2018-07-31">0.3.2</a></li>
        <li><a href="#bcbiobase-0-3-1-2018-07-24">0.3.1</a></li>
        <li><a href="#bcbiobase-0-3-0-2018-07-17">0.3.0</a></li>
        <li><a href="#bcbiobase-0-2-16-2018-06-28">0.2.16</a></li>
        <li><a href="#bcbiobase-0-2-15-2018-06-05">0.2.15</a></li>
        <li><a href="#bcbiobase-0-2-14-2018-05-18">0.2.14</a></li>
        <li><a href="#bcbiobase-0-2-13-2018-05-07">0.2.13</a></li>
        <li><a href="#bcbiobase-0-2-12-2018-05-08">0.2.12</a></li>
        <li><a href="#bcbiobase-0-2-11-2018-05-07">0.2.11</a></li>
        <li><a href="#bcbiobase-0-2-10-2018-05-03">0.2.10</a></li>
        <li><a href="#bcbiobase-0-2-9-2018-04-30">0.2.9</a></li>
        <li><a href="#bcbiobase-0-2-8-2018-04-25">0.2.8</a></li>
        <li><a href="#bcbiobase-0-2-7-2018-04-25">0.2.7</a></li>
        <li><a href="#bcbiobase-0-2-6-2018-04-25">0.2.6</a></li>
        <li><a href="#bcbiobase-0-2-5-2018-04-24">0.2.5</a></li>
        <li><a href="#bcbiobase-0-2-4-2018-04-22">0.2.4</a></li>
        <li><a href="#bcbiobase-0-2-3-2018-04-19">0.2.3</a></li>
        <li><a href="#bcbiobase-0-2-2-2018-04-16">0.2.2</a></li>
        <li><a href="#bcbiobase-0-2-1-2018-04-13">0.2.1</a></li>
        <li><a href="#bcbiobase-0-2-0-2018-03-22">0.2.0</a></li>
        <li><a href="#bcbiobase-0-1-4-2018-02-20">0.1.4</a></li>
        <li><a href="#bcbiobase-0-1-3-2018-02-19">0.1.3</a></li>
        <li><a href="#bcbiobase-0-1-2-2018-02-18">0.1.2</a></li>
        <li><a href="#bcbiobase-0-1-1-2018-02-17">0.1.1</a></li>
        <li><a href="#bcbiobase-0-1-0-2018-02-13">0.1.0</a></li>
        <li><a href="#bcbiobase-0-0-3-2018-01-27">0.0.3</a></li>
        <li><a href="#bcbiobase-0-0-2-2018-01-19">0.0.2</a></li>
        <li><a href="#bcbiobase-0-0-1-2018-01-11">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

